spec/arrays.dd

```diff
@@ -759,24 +760,63 @@ array.length = i;
 
 $(H3 $(LNAME2 capacity-reserve, `capacity` and `reserve`))
 
-        $(P The $(D capacity) property gives the maximum length the array
-        can grow to without reallocating. The spare capacity for array *a*
-        is `a.capacity - a.length`.)
+        $(P The $(D capacity) property gives the maximum length a dynamic array
+        can grow to without reallocating. If the array does not point to
+        GC-allocated memory, the capacity will be zero.
+        The spare capacity for an array *a* is `a.capacity - a.length`.)
+
         $(P The $(D reserve)
-        function expands an array's capacity for use by the append operator.)
+        function expands an array's capacity for use by the
+        $(RELATIVE_LINK2 array-appending, append operator) or `.length` assignment.)
 
-$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+$(SPEC_RUNNABLE_EXAMPLE_RUN
 ---------
 int[] array;
-size_t cap = array.reserve(10); // request
+assert(array.capacity == 0);
+const size_t cap = array.reserve(10); // request
 assert(cap >= 10); // allocated may be more than request
+assert(array.ptr != null);
 
 int[] copy = array;
+assert(copy.capacity == cap); // array and copy have same capacity
 array ~= [1, 2, 3, 4, 5]; // grow in place
-assert(cap == array.capacity);
+assert(cap == array.capacity); // array memory was not reallocated
 assert(copy.ptr == array.ptr);
+assert(copy.capacity == 0);
+copy ~= 0; // new allocation
+assert(copy.ptr != array.ptr);
 ---------
 )
+        $(P Above, `copy`'s length remains zero but it points to the same
+        memory allocated by the `reserve` call. Because `array` is then appended
+        to, `copy.ptr + 0` no longer points to unused memory - instead that
+        is the address of `array[0]`. So `copy.capacity` will be zero to
+        prevent any appending to `copy` from overwriting elements in `array`.)
+
+        $(NOTE The runtime uses the number of elements appended to track the
+        start of the spare capacity for the memory allocation.)
+
+        $(P When an array with spare capacity has its length reduced, or is
+        assigned a slice of itself that ends before the previous last element,
+        the capacity will be zero.)
+
+        $(SPEC_RUNNABLE_EXAMPLE_RUN
+        ---
+        int[] a = [1, 2, 3];
+        a.reserve(2);
+        a = a[];
+        assert(a.capacity > 0);
+        a = a[0..2];
+        assert(a.capacity == 0);
+        ---
+        )
+
+        $(NOTE Accessing `.capacity` requires the runtime to
+        acquire a global lock and perform a table lookup.)
```

Comment:
This is not true. There is a thread-local cache for capacity/appending. The lock is only required to fill in the cache, and then just the cache is used.

However, it still can have a performance impact, because you are talking about an opaque function that cannot be inlined, and the cache lookup isn't really that cheap either.

Reply:
OK, I've changed this to `may require`.